### PR TITLE
[BEAM-822] Adjust how and when we build package-info.java to avoid spurious rebuilds

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Together with #1204 this will eliminate spurious rebuilds of the Java SDK.

R: @lukecwik  and @swegner does this violate anything you had in mind for `build-tools`? It could also go in the core SDK, I suppose. It is nice to put it somewhere that users will not depend on it. (so if they obey IWYU it is not part of the API)

### CODE REVIEW

**Comment by swegner** [comment]
What a mess. Thanks for tackling this, and for linking to relevant context. If I understand correctly: 

1. `package-info.java` files generate spurious rebuilds because they don't by default generate a `.class` file which can be used to track necessary rebuild.
2. Adding a package annotation in the file causes it to generate a `.class` file.
3. The only standard annotation in JDK 1.5 is `@Deprecated`, so instead this defines a `@Package` annotation.

Did I grok that correctly? Makes sense to me. 

[One comment](https://issues.apache.org/jira/browse/MCOMPILER-205?focusedCommentId=14428953&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14428953) recommends `<exclude>`'ing the problematic files from incremental analysis. Have you tried that? This would presumably only be a problem if you made local changes to a `package-info.java` file which were important for rebuild. Since ours are all boilerplate, this shouldn't be an issue.

re: Is this ok for `build-tools`: I imagine `build-tools` for build resources rather than code dependencies, but that doesn't need be the case. This `@Package` annotation is specifically for influencing the build process, so in that context `build-tools` is the right place.

**Comment by lukecwik** [comment]
Isn't this a problem for all java modules and not just sdks/java/core?

The reason why I don't like this annotation based approach is that it requires all package info files to contain it to prevent this problem and is to workaround a limitation in the maven compiler plugin. Which is why I would prefer if we configured the maven compiler plugin in the root pom.xml to use -Xpkginfo:always or have a configuration with an exclusion line which filters out **/package-info.java?

**Comment by kennknowles** [comment]
@swegner yes, you got it. With the one missing piece that `-Xpkginfo:always` was added in Java 7 but in fact _always_ throws a NPE in all Java 7 JDKs we use :-). I previously proposed a PR with `<exclude>` but @davorbonaci didn't like that. I do think that changes to the javadoc might get missed? Surely a release build always starts with a `mvn clean` but just to be safe, I do agree with him that it is nice to not exclude them. I'm happy either way.

@lukecwik it could be, but the SDK is the one that I noticed having the rebuilds, while other modules did not. I didn't look into the status of other modules. I know our javadoc and/or checkstyle config now insists on having them. I'm willing to tag the rest here, but I prefer to quickly merge anything that is an improvement on status quo.

**Comment by lukecwik** [comment]
I'm not a big fan of adding something which most people will not know about or check for making it easy to break what this change is attempting to do.

Good to know about -Xpkginfo not working. Any idea about just using the exclusion on **/package-inf.java?

**Comment by kennknowles** [comment]
Apparently the JDK bug is only triggered when there is _also_ actual Java code. That might explain why it snuck past their testing.

Now this is split into two executions, so the `package-info.java` compilation can use `-Xpkginfo:always`. This is, I believe, a nearly ideal solution. I have confirm that merging #1204 and #1205 into master suppressed spurious rebuilds.

**Comment by kennknowles** [comment]
Travis failures are Eclipse issues and Travis infra failure, respectively. There was a non-crashing execution on Java 7, which is the important thing for the tests as far as this PR goes. Jenkins adds no coverage, since it is on a newer Java.

File: pom.xml [hunk 1]
```diff
@@ -831,6 +831,44 @@
             <!-- Another temp override, to be set to true in due course. -->
             <showDeprecation>false</showDeprecation>
           </configuration>
+          <executions>
+
```
Comment (lukecwik) [inline pom.xml]: nit: remove single line of whitespace

File: pom.xml [hunk 2]
```diff
@@ -831,6 +831,44 @@
...
+                -->
+                <excludes>
+                  <exclude>**/package-info.java</exclude>
+                </excludes>
+              </configuration>
+            </execution>
+
+            <!-- 
+              Compile just package-info.java to avoid 
+              https://bugs.openjdk.java.net/browse/JDK-8022161 -->
```
Comment (lukecwik) [inline pom.xml]: nit: move --> to next line

File: pom.xml (hunk 1 above)
Comment (kennknowles) [inline pom.xml]: Reformatted slightly so the whitespace makes more sense.

File: pom.xml (hunk 2 above)
Comment (kennknowles) [inline pom.xml]: Done.

**Comment by dkulp** [comment]
FYI:  This completely breaks building within Eclipse, not sure what can be done to fix it yet.

**Comment by kennknowles** [comment]
Thanks for noticing this and for the workaround in #1265.

**Pull request closed without merging**